src/metabase/lib/js/metadata.cljs

```diff
@@ -0,0 +1,150 @@
+  [_metadata-type obj]
+  (obj->clj {:lib/type :metadata/table}
+            (comp (remove (fn [[k _v]]
+                            (= k "fields")))
+                  map-pair-value->clj-xform)
+            obj))
+
+(defmethod metadata->clj* :field
+  [_metadata-type obj]
+  (obj->clj {:lib/type :metadata/field}
+            (comp (remove (fn [[k _v]]
+                            (= k "table")))
+                  map-pair-value->clj-xform)
+            obj))
+
+(defmethod metadata->clj* :metric
+  [_metadata-type obj]
+  (obj->clj {:lib/type :metadata/metric}
+            (comp (remove (fn [[k _v]]
+                            (or (= k "database")
+                                (= k "table"))))
+                  map-pair-value->clj-xform)
+            obj))
+
+(defmethod metadata->clj* :segment
+  [_metadata-type obj]
+  (obj->clj {:lib/type :metadata/segment}
+            (comp (remove (fn [[k _v]]
+                            (or (= k "database")
+                                (= k "table"))))
+                  map-pair-value->clj-xform)
+            obj))
+
+(defn- metadata-map->clj [metadata-type m]
+  (into {}
+        (comp (map (fn [k]
+                     [(parse-long k) (gobject/get m k)]))
+              (remove (fn [[_k v]]
+                        (= (goog/typeOf v) "function")))
+              (map (fn [[k v]]
+                     [k (metadata->clj* metadata-type v)])))
+        (gobject/getKeys m)))
+
+(defn- metadata->clj [metadata]
+  (into {}
+        [(when-let [databases (gobject/get metadata "databases")]
+           [:databases (metadata-map->clj :database databases)])
+         (when-let [tables (gobject/get metadata "tables")]
+           [:tables (metadata-map->clj :table tables)])
+         (when-let [fields (gobject/get metadata "fields")]
+           [:fields (metadata-map->clj :field fields)])
+         (when-let [metrics (gobject/get metadata "metrics")]
+           [:metrics (metadata-map->clj :metric metrics)])
+         (when-let [segments (gobject/get metadata "segments")]
+           [:segments (metadata-map->clj :segment segments)])]))
+
+(defn- database [metadata database-id]
+  (get-in metadata [:databases database-id]))
+
+(defn- table [metadata table-id]
+  (get-in metadata [:tables table-id]))
+
+(defn- field [metadata field-id]
+  (get-in metadata [:fields field-id]))
+
+(defn- metric [metadata metric-id]
+  (get-in metadata [:metrics metric-id]))
+
+(defn- segment [metadata segment-id]
+  (get-in metadata [:segments segment-id]))
+
+(defn- tables [database-id metadata]
+  (filter (fn [table-metadata]
+            (= (:db_id table-metadata) database-id))
+          (get-in metadata [:tables])))
+
+(defn- fields [metadata table-id]
+  (filter (fn [table-metadata]
+            (= (:table_id table-metadata) table-id))
+          (get-in metadata [:fields])))
+
+(defrecord ^:private MLv1QuestionMetadataProvider [database-id metadata]
+  lib.metadata.protocols/MetadataProvider
```

Comment:
I overhauled the `DatabseMetadataProvider` protocol and added a handful of methods to it. We needed addition methods for metrics and segments. I added some methods to fetch specific things by ID, which will let us write implementations that are a little more efficient e.g. for QP/the QP store in the near-term future